src/erc721/module.cairo

```diff
@@ -0,0 +1,250 @@
+    }
+
+    #[derive(Drop, starknet::Event)]
+    struct Approval {
+        owner: ContractAddress,
+        approved: ContractAddress,
+        token_id: u256,
+    }
+
+    #[derive(Drop, starknet::Event)]
+    struct ApprovalForAll {
+        owner: ContractAddress,
+        operator: ContractAddress,
+        approved: bool,
+    }
+
+    #[constructor]
+    fn constructor(ref self: ContractState) {
+        self.initializer();
+    }
+
+    #[external(v0)]
+    impl ERC721Impl of IERC721<ContractState> {
+        fn balance_of(self: @ContractState, account: ContractAddress) -> u256 {
+            assert(!account.is_zero(), 'ERC721: invalid account');
+            self._balances.read(account)
+        }
+
+        fn owner_of(self: @ContractState, token_id: u256) -> ContractAddress {
+            self._owner_of(token_id)
+        }
+
+        fn get_approved(self: @ContractState, token_id: u256) -> ContractAddress {
+            assert(self._exists(token_id), 'ERC721: invalid token ID');
+            self._token_approvals.read(token_id)
+        }
+
+        fn is_approved_for_all(
+            self: @ContractState, owner: ContractAddress, operator: ContractAddress
+        ) -> bool {
+            self._operator_approvals.read((owner, operator))
+        }
+
+        fn approve(ref self: ContractState, to: ContractAddress, token_id: u256) {
+            let owner = self._owner_of(token_id);
+
+            let caller = get_caller_address();
+            assert(
+                owner == caller || self.is_approved_for_all(owner, caller),
+                'ERC721: unauthorized caller'
+            );
+            self._approve(to, token_id);
+        }
+
+        fn set_approval_for_all(
+            ref self: ContractState, operator: ContractAddress, approved: bool
+        ) {
+            self._set_approval_for_all(get_caller_address(), operator, approved)
+        }
+
+        fn transfer_from(
+            ref self: ContractState, from: ContractAddress, to: ContractAddress, token_id: u256
+        ) {
+            assert(
+                self._is_approved_or_owner(get_caller_address(), token_id),
+                'ERC721: unauthorized caller'
+            );
+            self._transfer(from, to, token_id);
+        }
+
+        fn safe_transfer_from(
+            ref self: ContractState,
+            from: ContractAddress,
+            to: ContractAddress,
+            token_id: u256,
+            data: Span<felt252>
+        ) {
+            assert(
+                self._is_approved_or_owner(get_caller_address(), token_id),
+                'ERC721: unauthorized caller'
+            );
+            self._safe_transfer(from, to, token_id, data);
+        }
+    }
+
+    #[generate_trait]
+    impl InternalImpl of InternalTrait {
+        fn initializer(ref self: ContractState) {
+            let mut unsafe_state = SRC5::unsafe_new_contract_state();
+            SRC5::InternalImpl::register_interface(ref unsafe_state, IERC721_ID);
```

Comment:
Do we need to use `register_interface` here or can/should we hardcode them in order to save storage?

Reply:
We could indeed overload the `supports_interface` SRC5 method but I think this is a real hack of how it must be used